Lambda School Data Science

*Unit 2, Sprint 1, Module 2*

---

# Regression 2

## Assignment

You'll continue to **predict how much it costs to rent an apartment in NYC,** using the dataset from renthop.com.

- [ ] Do train/test split. Use data from April & May 2016 to train. Use data from June 2016 to test.
- [ ] Engineer at least two new features. (See below for explanation & ideas.)
- [ ] Fit a linear regression model with at least two features.
- [ ] Get the model's coefficients and intercept.
- [ ] Get regression metrics RMSE, MAE, and $R^2$, for both the train and test data.
- [ ] What's the best test MAE you can get? Share your score and features used with your cohort on Slack!
- [ ] As always, commit your notebook to your fork of the GitHub repo.


#### [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)

> "Some machine learning projects succeed and some fail. What makes the difference? Easily the most important factor is the features used." — Pedro Domingos, ["A Few Useful Things to Know about Machine Learning"](https://homes.cs.washington.edu/~pedrod/papers/cacm12.pdf)

> "Coming up with features is difficult, time-consuming, requires expert knowledge. 'Applied machine learning' is basically feature engineering." — Andrew Ng, [Machine Learning and AI via Brain simulations](https://forum.stanford.edu/events/2011/2011slides/plenary/2011plenaryNg.pdf) 

> Feature engineering is the process of using domain knowledge of the data to create features that make machine learning algorithms work. 

#### Feature Ideas
- Does the apartment have a description?
- How long is the description?
- How many total perks does each apartment have?
- Are cats _or_ dogs allowed?
- Are cats _and_ dogs allowed?
- Total number of rooms (beds + baths)
- Ratio of beds to baths
- What's the neighborhood, based on address or latitude & longitude?

## Stretch Goals
- [ ] If you want more math, skim [_An Introduction to Statistical Learning_](http://faculty.marshall.usc.edu/gareth-james/ISL/ISLR%20Seventh%20Printing.pdf),  Chapter 3.1, Simple Linear Regression, & Chapter 3.2, Multiple Linear Regression
- [ ] If you want more introduction, watch [Brandon Foltz, Statistics 101: Simple Linear Regression](https://www.youtube.com/watch?v=ZkjP5RJLQF4)
(20 minutes, over 1 million views)
- [ ] Add your own stretch goal(s) !

In [57]:
# %%capture
# import sys

# If you're on Colab:
# if 'google.colab' in sys.modules:
#     DATA_PATH = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Applied-Modeling/master/data/'
#     !pip install category_encoders==2.*

# If you're working locally:
# else:
    # DATA_PATH = '../data/'

# my data
# since I'm using my local conda setting
DATA_PATH = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Applied-Modeling/master/data/'

# Ignore this Numpy warning when using Plotly Express:
# FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
import warnings
warnings.filterwarnings(action='ignore', category=FutureWarning, module='numpy')

In [69]:
import numpy as np
import pandas as pd

# Read New York City apartment rental listing data
df = pd.read_csv(DATA_PATH+'apartments/renthop-nyc.csv', na_filter=False)
assert df.shape == (49352, 34)

# Remove the most extreme 1% prices,
# the most extreme .1% latitudes, &
# the most extreme .1% longitudes
df = df[(df['price'] >= np.percentile(df['price'], 0.5)) & 
        (df['price'] <= np.percentile(df['price'], 99.5)) & 
        (df['latitude'] >= np.percentile(df['latitude'], 0.05)) & 
        (df['latitude'] < np.percentile(df['latitude'], 99.95)) &
        (df['longitude'] >= np.percentile(df['longitude'], 0.05)) & 
        (df['longitude'] <= np.percentile(df['longitude'], 99.95))]

In [70]:
df_backup = df

In [71]:
# run this if you want to restart
df = df_backup

In [67]:
# check data samples
df.head()

bathrooms  bedrooms              created  \
0        1.5         3  2016-06-24 07:54:24   
1        1.0         2  2016-06-12 12:19:27   
2        1.0         1  2016-04-17 03:26:41   
3        1.0         1  2016-04-18 02:22:02   
4        1.0         4  2016-04-28 01:32:41   

                                         description      display_address  \
0  A Brand New 3 Bedroom 1.5 bath ApartmentEnjoy ...  Metropolitan Avenue   
1                                                         Columbus Avenue   
2  Top Top West Village location, beautiful Pre-w...          W 13 Street   
3  Building Amenities - Garage - Garden - fitness...     East 49th Street   
4  Beautifully renovated 3 bedroom flex 4 bedroom...    West 143rd Street   

   latitude  longitude  price           street_address interest_level  ...  \
0   40.7145   -73.9425   3000  792 Metropolitan Avenue         medium  ...   
1   40.7947   -73.9667   5465      808 Columbus Avenue            low  ...   
2   40.7388   -74.0018   2850          241 W 13 Street           high  ...   
3   40.7539   -73.9677   3275     333 East 49th Street            low  ...   
4   40.8241   -73.9493   3350    500 West 143rd Street            low  ...   

   high_speed_internet  balcony  swimming_pool  new_construction  terrace  \
0                    0        0              0                 0        0   
1                    0        0              0                 0        0   
2                    0        0              0                 0        0   
3                    0        0              0                 0        0   
4                    0        0              0                 0        0   

   exclusive  loft  garden_patio  wheelchair_access  common_outdoor_space  
0          0     0             0                  0                     0  
1          0     0             0                  0                     0  
2          0     0             0                  0                     0  
3          0     0             0                  0                     0  
4          0     0             0                  0                     0  

[5 rows x 34 columns]

In [5]:
df.dtypes

bathrooms               float64
bedrooms                  int64
created                  object
description              object
display_address          object
latitude                float64
longitude               float64
price                     int64
street_address           object
interest_level           object
elevator                  int64
cats_allowed              int64
hardwood_floors           int64
dogs_allowed              int64
doorman                   int64
dishwasher                int64
no_fee                    int64
laundry_in_building       int64
fitness_center            int64
pre-war                   int64
laundry_in_unit           int64
roof_deck                 int64
outdoor_space             int64
dining_room               int64
high_speed_internet       int64
balcony                   int64
swimming_pool             int64
new_construction          int64
terrace                   int64
exclusive                 int64
loft                      int64
garden_p

# Assignment again
1. Do train/test split. Use data from April & May 2016 to train. Use data from June 2016 to test.
2. Engineer at least two new features. (See below for explanation & ideas.)
3. Fit a linear regression model with at least two features.
4. Get the model's coefficients and intercept.
5. Get regression metrics RMSE, MAE, and R^2, for both the train and test data.
6j. What's the best test MAE you can get? Share your score and features used with your cohort on Slack!

As always, commit your notebook to your fork of the GitHub repo.

In [72]:
# quick data check
df['created'].head()

0    2016-06-24 07:54:24
1    2016-06-12 12:19:27
2    2016-04-17 03:26:41
3    2016-04-18 02:22:02
4    2016-04-28 01:32:41
Name: created, dtype: object

In [73]:
# check created date in ascending order
df['created'].sort_values(ascending=True)

5186     2016-04-01 22:12:41
7945     2016-04-01 22:56:00
6424     2016-04-01 22:57:15
7719     2016-04-01 23:26:07
1723     2016-04-02 00:48:13
                ...         
19943    2016-06-29 17:47:34
16801    2016-06-29 17:56:12
32633    2016-06-29 18:14:48
20560    2016-06-29 18:30:41
17743    2016-06-29 21:41:47
Name: created, Length: 48817, dtype: object

In [74]:
# 1. do train/test split
condition_train = (df['created'] > '2016-04-00 00:00:00') & (df['created'] < '2016-06-00 00:00:00')
condition_test = df['created'] > '2016-06-00 00:00:00'
train = df[condition_train]
test = df[condition_test]
print('train', train.shape)
print('test', test.shape)


train (31844, 34)
test (16973, 34)


# 2. Feature Engineering

Feature Ideas

- Does the apartment have a description?
- How long is the description?
- How many total perks does each apartment have?
- Are cats or dogs allowed?
- Are cats and dogs allowed?
- Total number of rooms (beds + baths)
- Ratio of beds to baths
- What's the neighborhood, based on address or latitude & longitude?

In [43]:
# again dtypes
df.dtypes

bathrooms               float64
bedrooms                  int64
created                  object
description              object
display_address          object
latitude                float64
longitude               float64
price                     int64
street_address           object
interest_level           object
elevator                  int64
cats_allowed              int64
hardwood_floors           int64
dogs_allowed              int64
doorman                   int64
dishwasher                int64
no_fee                    int64
laundry_in_building       int64
fitness_center            int64
pre-war                   int64
laundry_in_unit           int64
roof_deck                 int64
outdoor_space             int64
dining_room               int64
high_speed_internet       int64
balcony                   int64
swimming_pool             int64
new_construction          int64
terrace                   int64
exclusive                 int64
loft                      int64
garden_p

In [94]:
df['description'][31827]
# we need to figure some rows has no string in values

'        '

In [93]:
# create a feature has_desc boolean
df['has_desc'] = df['description'].apply(lambda x: 0 if x.isspace() else 1)

In [92]:
# check if has_desc is updated
df['has_desc'].sample(20)

27263    1
28808    1
24459    1
38214    1
37       1
31478    1
9892     1
38422    1
34438    1
22949    1
12714    1
27160    1
10819    1
2142     1
48758    1
45793    1
48486    1
42412    0
27815    1
41767    1
Name: has_desc, dtype: int64

# features to select
well let me choose lantitude and longitude as my features
That means I can choose selecting features and we don't need feature engineering
my assumption is that rental price varies based on latitude & longitutde combinations because it determines the neighborhood in New York

In [95]:
import plotly.express as px

In [105]:
# 1. Import the appropriate estimator class from sk learn
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [101]:
# 2. Instantiate this class
model = LinearRegression()
model

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [103]:
# 3. Arrange X features matrices 
features = ['latitude', 'longitude']
X_train = train[features]
x_test = test[features]

In [104]:
# 4. Fit the model
target = 'price'
y_train = train[target]
model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [106]:
# 5. Apply the model to the new data
y_pred_train = model.predict(X_train)
mean_absolute_error(y_pred_train, y_train)

1147.149327823136

In [108]:
y_pred = model.predict(x_test)
y_test = test[target]
mean_absolute_error(y_pred, y_test)

1139.7004576307802

I have `1147` for train data mean abs error
and `1139` for test data mean abs error

In [109]:
import itertools
import numpy as np
import plotly.express as px
import plotly.graph_objs as go
from sklearn.linear_model import LinearRegression

def regression_3d(df, x, y, z, num=100, **kwargs):
    """
    Visualize linear regression in 3D: 2 features + 1 target
    
    df : Pandas DataFrame
    x : string, feature 1 column in df
    y : string, feature 2 column in df
    z : string, target column in df
    num : integer, number of quantiles for each feature
    """
    
    # Plot data
    fig = px.scatter_3d(df, x, y, z, **kwargs)
    
    # Fit Linear Regression
    features = [x, y]
    target = z
    model = LinearRegression()
    model.fit(df[features], df[target])    
    
    # Define grid of coordinates in the feature space
    xmin, xmax = df[x].min(), df[x].max()
    ymin, ymax = df[y].min(), df[y].max()
    xcoords = np.linspace(xmin, xmax, num)
    ycoords = np.linspace(ymin, ymax, num)
    coords = list(itertools.product(xcoords, ycoords))
    
    # Make predictions for the grid
    predictions = model.predict(coords)
    Z = predictions.reshape(num, num).T
    
    # Plot predictions as a 3D surface (plane)
    fig.add_trace(go.Surface(x=xcoords, y=ycoords, z=Z))
    
    return fig

In [110]:
regression_3d(
    train,
    x='latitude', 
    y='longitude', 
    z='price', 
    title='nighborhood vs price'
)

In [111]:
# get model intercept and coefficient
model.intercept_, model.coef_

(-1285931.9851655476, array([  2208.18971896, -16215.70541389]))

In [112]:
# This is easier to read
print('Intercept', model.intercept_)
coefficients = pd.Series(model.coef_, features)
print(coefficients.to_string())

Intercept -1285931.9851655476
latitude      2208.189719
longitude   -16215.705414


In [117]:
from matplotlib.patches import Rectangle
import matplotlib.pyplot as plt
import numpy as np
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

def squared_errors(df, feature, target, m, b):
    """
    Visualize linear regression, with squared errors,
    in 2D: 1 feature + 1 target.
    
    Use the m & b parameters to "fit the model" manually.
    
    df : Pandas DataFrame
    feature : string, feature column in df
    target : string, target column in df
    m : numeric, slope for linear equation
    b : numeric, intercept for linear requation
    """
    
    # Plot data
    # fig = plt.figure(figsize=(7,7))
    # ax = plt.axes()
    # df.plot.scatter(feature, target, ax=ax)
    
    # Make predictions
    x = df[feature]
    y = df[target]
    y_pred = m*x + b
    
    # Plot predictions
    ax.plot(x, y_pred)
    
    # Plot squared errors
    xmin, xmax = ax.get_xlim()
    ymin, ymax = ax.get_ylim()
    scale = (xmax-xmin)/(ymax-ymin)
    for x, y1, y2 in zip(x, y, y_pred):
        bottom_left = (x, min(y1, y2))
        height = abs(y1 - y2)
        width = height * scale
        ax.add_patch(Rectangle(xy=bottom_left, width=width, height=height, alpha=0.1))
    
    # Print regression metrics
    mse = mean_squared_error(y, y_pred)
    rmse = np.sqrt(mse)
    mae = mean_absolute_error(y, y_pred)
    r2 = r2_score(y, y_pred)
    print('Mean Squared Error:', mse)
    print('Root Mean Squared Error:', rmse)
    print('Mean Absolute Error:', mae)
    print('R^2:', r2)

In [118]:
feature = ['latitude','longitude']
squared_errors(train, feature, target, m=0, b=y_train.mean())

NameError: name 'ax' is not defined